src/gf.c

```diff
@@ -1039,9 +1065,67 @@ jl_lambda_info_t *jl_get_specialization1(jl_tupletype_t *types)
     return NULL;
 }
 
+// Specialize, but throw an error if there is a call ambiguity
+jl_lambda_info_t *jl_checked_specialization1(jl_tupletype_t *types)
+{
+    jl_typemap_entry_t *entry;
+    jl_lambda_info_t *linfo = jl_get_specialization1(types, &entry);
+    if (linfo != NULL) {
+        jl_method_t *m = linfo->def;
+        jl_value_t *ambiguities = jl_list_call_ambiguities_(types, m);
+        if (ambiguities != jl_nothing) {
+            jl_ambiguous_method_error((jl_value_t*)types, ambiguities);
+        }
+    }
+    return linfo;
+}
+
 JL_DLLEXPORT void jl_compile_hint(jl_tupletype_t *types)
 {
-    (void)jl_get_specialization1(types);
+    (void)jl_checked_specialization1(types);
```

Comment:
I think this function should continue not to throw errors. Also, the callers of `jl_get_specialization1` expect it to return NULL if anything goes wrong instead of throwing (if any callers don't check for NULL, they should). Particularly in codegen.cpp we should avoid throwing exceptions.